mpcs/src/util/merkle_tree.rs

```diff
@@ -23,36 +23,89 @@ use ark_std::{end_timer, start_timer};
 use super::hash::write_digest_to_transcript;
 
 #[derive(Clone, Debug, Default, Serialize, Deserialize)]
-#[serde(bound(serialize = "E: Serialize", deserialize = "E: DeserializeOwned"))]
-pub struct MerkleTree<E: ExtensionField>
+#[serde(bound(deserialize = "E: DeserializeOwned"))]
+pub struct MerkleTreeDigests<E: ExtensionField>
 where
     E::BaseField: Serialize + DeserializeOwned,
 {
     inner: Vec<Vec<Digest<E::BaseField>>>,
-    leaves: Vec<FieldType<E>>,
 }
 
-impl<E: ExtensionField> MerkleTree<E>
+impl<E: ExtensionField> MerkleTreeDigests<E>
 where
     E::BaseField: Serialize + DeserializeOwned,
 {
-    pub fn compute_inner(leaves: &FieldType<E>) -> Vec<Vec<Digest<E::BaseField>>> {
-        merkelize::<E>(&[leaves])
+    pub fn from_leaves(leaves: &FieldType<E>) -> Self {
+        Self {
+            inner: merkelize::<E>(&[leaves]),
+        }
+    }
+    pub fn from_leaves_base(leaves: &[E::BaseField]) -> Self {
+        Self {
+            inner: merkelize_base::<E>(&[leaves]),
+        }
+    }
+    pub fn from_leaves_ext(leaves: &[E]) -> Self {
+        Self {
+            inner: merkelize_ext::<E>(&[leaves]),
+        }
     }
 
-    pub fn compute_inner_base(leaves: &[E::BaseField]) -> Vec<Vec<Digest<E::BaseField>>> {
-        merkelize_base::<E>(&[leaves])
+    pub fn from_batch_leaves(leaves: &[&FieldType<E>]) -> Self {
+        Self {
+            inner: merkelize::<E>(leaves),
+        }
     }
 
-    pub fn compute_inner_ext(leaves: &[E]) -> Vec<Vec<Digest<E::BaseField>>> {
-        merkelize_ext::<E>(&[leaves])
+    pub fn root(&self) -> Digest<E::BaseField> {
+        self.inner.last().unwrap()[0].clone()
     }
 
-    pub fn root_from_inner(inner: &[Vec<Digest<E::BaseField>>]) -> Digest<E::BaseField> {
-        inner.last().unwrap()[0].clone()
+    pub fn root_ref(&self) -> &Digest<E::BaseField> {
+        &self.inner.last().unwrap()[0]
     }
 
-    pub fn from_inner_leaves(inner: Vec<Vec<Digest<E::BaseField>>>, leaves: FieldType<E>) -> Self {
+    pub fn height(&self) -> usize {
+        self.inner.len()
+    }
+
+    pub fn bottom_size(&self) -> usize {
+        self.inner.first().unwrap().len()
+    }
+
+    pub fn merkle_path_without_leaf_sibling_or_root(
+        &self,
+        leaf_group_index: usize, // Two leaves make a group.
```

Comment:
Perhaps stick that in a doc comment instead, so it shows up when people try to use this function?